dimod/include/dimod/quadratic_model.h

```diff
@@ -322,8 +229,233 @@ class Neighborhood {
+    using pointer = value_type*;
+    using reference = value_type&;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category = std::random_access_iterator_tag;
+
+    NeighborhoodIterator() : neighborhood_(nullptr), i_(0), pair_(nullptr) {}
+
+    NeighborhoodIterator(neighborhood_type* neighborhood, index_type i)
+            : neighborhood_(neighborhood), i_(i), pair_(nullptr) {
+        update_pair();
+    }
+
+    reference operator*() { return *pair_; }
+
+    pointer operator->() { return pair_; }
+
+    NeighborhoodIterator<Bias, Index>& operator++() {
+        ++i_;
+        update_pair();
+        return *this;
+    }
+    NeighborhoodIterator<Bias, Index>& operator--() {
+        --i_;
+        update_pair();
+        return *this;
+    }
+    NeighborhoodIterator<Bias, Index> operator++(int) {
+        NeighborhoodIterator<Bias, Index> r(*this);
+        ++i_;
+        update_pair();
+        return r;
+    }
+    NeighborhoodIterator<Bias, Index> operator--(int) {
+        NeighborhoodIterator<Bias, Index> r(*this);
+        --i_;
+        update_pair();
+        return r;
+    }
+
+    NeighborhoodIterator<Bias, Index>& operator+=(int n) {
+        i_ += n;
+        update_pair();
+        return *this;
+    }
+    NeighborhoodIterator<Bias, Index>& operator-=(int n) {
+        i_ -= n;
+        update_pair();
+        return *this;
+    }
+
+    NeighborhoodIterator<Bias, Index> operator+(int n) const {
+        NeighborhoodIterator<Bias, Index> r(*this);
+        return r += n;
+    }
+    NeighborhoodIterator<Bias, Index> operator-(int n) const {
+        NeighborhoodIterator<Bias, Index> r(*this);
+        return r -= n;
+    }
+
+    difference_type operator-(
+            NeighborhoodIterator<Bias, Index> const& r) const {
+        return i_ - r.i_;
+    }
+
+    bool operator<(NeighborhoodIterator<Bias, Index> const& r) const {
+        return i_ < r.i_;
+    }
+    bool operator<=(NeighborhoodIterator<Bias, Index> const& r) const {
+        return i_ <= r.i_;
+    }
+    bool operator>(NeighborhoodIterator<Bias, Index> const& r) const {
+        return i_ > r.i_;
+    }
+    bool operator>=(NeighborhoodIterator<Bias, Index> const& r) const {
+        return i_ >= r.i_;
+    }
+    bool operator!=(const NeighborhoodIterator<Bias, Index>& r) const {
+        return i_ != r.i_;
+    }
+    bool operator==(const NeighborhoodIterator<Bias, Index>& r) const {
+        return i_ == r.i_;
+    }
+
+ private:
+    neighborhood_type* neighborhood_;
+    index_type i_;
+
+    value_type* pair_;
+
+    void update_pair() {
+        if (pair_) {
+            delete pair_;
+        }
+        if (i_ < (index_type)neighborhood_->size()) {
+            pair_ = new value_type(neighborhood_->neighbors[i_],
+                                   neighborhood_->quadratic_biases[i_]);
+        }
+    }
+};
+
+template <class Bias, class Index>
+class ConstNeighborhoodIterator {
```

Comment:
can you do something like this ? 

template<clas Bias, class Index> 
BaseNeighborhoodIterator  -> define only this one. 

template<class Bias, class Index>
using NeighborhoodIterator = BaseNeighborhoodIterator<Bias, Index>

template <class Bias, class Index> 
using ConstNeighborhoodIteraotr = BaseNeighborhoodIterator<const Bias, Index>

Reply:
I don't think we could mark `const_neighborhood_iterator cbegin() const {}` in that case, since both would need to accept a non-const pointer.

I do totally hate the redundancy but as far as I tell you have to duplicate... Yay c++.